File: openrest4net/openrest4net-api/com/googlecode/openrest/Variation.cs
using System;
using System.Collections.Generic;

namespace com.googlecode.openrest
{
    /**
     * A possible modification to an item, e.g. "choice of sides" for a happy-meal
     * or "degree of cooking" for a hamburger.
     * @author DL
     */
    public class Variation
    {
        public const string VARIATION_DISPLAY_TYPE_DIFF = "diff";
        public const string VARIATION_DISPLAY_TYPE_CHOICE = "choice";

        public Variation(string title, string tagId, int minNumAllowed, int maxNumAllowed,
            IDictionary<string, int> prices, IList<string> defaults, string displayType)
        {
            this.title = title;
            this.tagId = tagId;
            this.minNumAllowed = minNumAllowed;
            this.maxNumAllowed = maxNumAllowed;
            this.prices = prices;
            this.defaults = defaults;
            this.displayType = displayType;
        }

        /** Empty constructor required for initialization from JSON-encoded string. */
        public Variation() { }

        /** The variations's name, e.g. "sides", "degree of cooking". */
        public string title;

        /** The set's name, e.g. "drink", "sides". */
        public string tagId;

        /** Minimum number of items to select. */
        public int? minNumAllowed = 0;

        /** Maximum number of items to select. */
        public int? maxNumAllowed = Int32.MaxValue;

        /** Items' base prices. Non-referenced items are free by default. */
        public IDictionary<string, int> prices = new Dictionary<string, int>();

        /** Default selected item ids. */
        public IList<string> defaults = new List<string>();

        /** Display type for human-readable printing. */
        public string displayType = VARIATION_DISPLAY_TYPE_CHOICE;
    }
}

File: trunk/openrest4js/openrest4js-api/Variation.js
OPENREST_VARIATION_DISPLAY_TYPE_DIFF = "diff";
OPENREST_VARIATION_DISPLAY_TYPE_CHOICE = "choice";
    
function VariationFromObj(data)
{
    return new Variation(data.title, data.itemIds, data.minNumAllowed, 
            data.maxNumAllowed, data.prices, data.defaults, data.displayType);
}

function Variation(title, itemIds, minNumAllowed, maxNumAllowed, prices, defaults, displayType) {
    this.title = title;
    this.itemIds = itemIds;
    this.minNumAllowed = minNumAllowed;
    this.maxNumAllowed = maxNumAllowed;
    this.prices = prices;
    this.defaults = defaults;
    this.displayType = displayType;
}

Variation.prototype.toOpenRestObj = function()
{
    var ret = {};
    
    if (typeof(this.title) != "undefined") ret['title'] = this.title;
    if (typeof(this.itemIds) != "undefined") ret['itemIds'] = this.itemIds;
    if (typeof(this.minNumAllowed) != "undefined") ret['minNumAllowed'] = this.minNumAllowed;
    if (typeof(this.maxNumAllowed) != "undefined") ret['maxNumAllowed'] = this.maxNumAllowed;
    if (typeof(this.prices) != "undefined") ret['prices'] = this.prices;
    if (typeof(this.defaults) != "undefined") ret['defaults'] = this.defaults;
    if (typeof(this.displayType) != "undefined") ret['displayType'] = this.displayType;

    return ret;
}
